### PR TITLE
[NO MERGE] TEST_INDEXING TRACKER

### PR DESCRIPTION
**Pull request by geohotstan**

### CODE REVIEW

**Comment by chenyuxyz** [comment]
@geohotstan if you are interested to fix all the getitem related test, i can make a bounty and lock to you "$200 pass all getitem tests in test_indexing.py"

**Comment by geohotstan** [comment]
image

interesting...

**Comment by chenyuxyz** [comment]
yeah torch does not support but we should, e.g. https://github.com/tinygrad/tinygrad/blob/5db0cdfbd3db5decb84127377ee8afad54274667/test/test_ops.py#L631

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -301,16 +301,29 @@ def pad(self, arg:Tuple[Optional[Tuple[int, int]], ...], value:float=0.0) -> Ten
   #        - if first Tensor passed in (expand dims) is not at dim 0
   #        - and following Tensors does not follow consecutively to the end of fancy indexing's dims
   def __getitem__(self, indices) -> Tensor: # indices: Union[int, slice, Tensor, None, Ellipsis, List, Tuple[Union[int, slice, Tensor, None, Ellipsis], ...]]
+    def validate_slice(sl: slice):
```
Comment (chenyuxyz) [inline tinygrad/tensor.py:304]: a general direction i want to split these and make it more manageable is (1) input normalization and validation (2) basic indexing, where we return a tensor with the same base as input (like a torch view), (3) advanced indexing, where we return a copy. think about this separation and add helper functions where you see fit

**Comment by geohotstan** [comment]
update:
ok I have an idea on how to do bool indexing.
It's a little hacked and may be a little controversial, but doing it this way will also allow better fancy indexing I think. I'll put up a proposal PR for the proof of concept real soon.

**Comment by chenyuxyz** [comment]
what's the status of this, any more feature you want to add? i think we have most of the important ones. want to clean up the document and close this one?

**Comment by geohotstan** [comment]
> what's the status of this, any more feature you want to add? i think we have most of the important ones. want to clean up the document and close this one?

Sry for the lack of updates, some immediate life related stuff came up and I'm on the road again.

I think I'm close to getting setitem working
if setitem works, I can make getitem and setitem share a lot of the code.
I also want to have getitem's fancy use parts of setitem's stuff. It's easier to reason about. (will have to check perf later tho once I get it correct)

**Comment by geohotstan** [comment]
The two PRs that'll complete this will be setitem (I'm just missing like a small piece of logic) and final test_indexing clean ups
I might do the final cleanups first
once that's done, I'll close this.

**Comment by chenyuxyz** [comment]
yea let's have the final cleanup first, i am ready to pay the bounty once that's done

**Pull request closed without merging**